runtime/src/comm/ofi/comm-ofi-oob-pmi2.c

```diff
@@ -286,6 +291,51 @@ void decode_kvs(char* raw, const char* enc, size_t size) {
 }
 
 int chpl_comm_ofi_oob_locales_on_node(void) {
-  // TODO: figure out how to implement this correctly
-  return 1;
+  int count = 0;
+  if (PMI_Get_clique_size != NULL) {
+    PMI_CHK(PMI_Get_clique_size(&count));
+    DBG_PRINTF(DBG_OOB, "PMI_Get_clique_size returned %d", count);  
+  } else {
+    // do an allgather of hostname hashes to determine the locales on the same node as us
+    // assumes each hostname has a unique hash
+
+    char hostname[256];
+    int rc = gethostname(hostname, sizeof(hostname));
+    CHK_TRUE(rc == 0);
+
+    uint64_t hash = 0;
+    for (int i = 0; i < sizeof(hostname); i++) {
+      char c = hostname[i];
+      if (c == '\0') {
+        break;
+      }
+      // The hash code is borrowed from gasnet including the comment.
+
+      /* The "c = ..." squeezes ASCII down to 6 bits, while encoding
+       * all chars valid in hostnames and IP addresses (IPV4 and IPV6).
+       * A unique value is assigned to each of the digits, the lower
+       * case letters, '-', '.' and ':'.  The upper case letters map
+       * to the same values as the corresponding lower-case.
+       */
+      c = ((c & 0x40) >> 1) | (c & 0x1f);
+      hash = ((hash << 6) | ((hash >> 58) & 0x3F)) ^ c;
+    }
```

Comment:
@PHHargrove @bonachea we based the above hashing of hostnames on `gasneti_hosthash`. It's a small amount of code, but I wanted to check that this and the attribution is ok, or if we need something more formal to indicate where this is from.

Reply:
Seems fine given the small amount of code.

If/when you borrow a larger piece of code or entire file, it would be preferable to include a reference to the third-party/gasnet/gasnet-src/license.txt file that contains full information.

Reply:
I'll add the reference. I wasn't sure how to attribute a small piece of code.